[MOB-294] BezierButton(SwiftUI) 추가

solchan87 opened the pull request:
## 어떤 PR 인가요?
<!-- 3줄이내로 PR에 대해 간략히 작성합니다. 3줄을 넘어야한다면 PR을 나누는걸 추천드려요. -->
BezierButton을 추가하는 PR 입니다.

## 작업 내용
<!-- 작업 리스팅, 리뷰어에게 필요한 작업 설명 등을 남겨주세요 -->
- SwiftUI 기반 BezierButton을 추가했습니다.
- SwiftUI 기반 BezierLoader 에서 부모 뷰의 애니메이션과 겹치는 문제로 UIKit BezierLoader를 추가하고 레핑해서 쓰는 방향으로 변경했습니다.
### 미비된 작업
- Pressed Color 논의가 진행중입니다.[팀챗](https://desk.channel.io/root/threads/groups/(CT)BezierRedesign-244958/66ab31c3600504251e46/66ac7300567869d4616d) 논의 후에 추가 PR로 작업 이어가겠습니다.

## 스크린샷 혹은 동영상
<!-- 
뷰가 변경된 경우에는 스크린샷을 첨부합니다. 
사진을 첨부한 뒤 src 링크를 변경해주세요. (적절히 width 조절)
사진이 여러장인 경우 테이블 가장 아래를 복사 / 붙여넣기합니다.
사진이 없는 경우 테이블을 제거합니다. 동영상인 경우는 테이블 바깥에 링크합니다.

ex) 동영상
https://user-images.githubusercontent.com/82797883/127593222-1d808786-2be1-4128-90ef-db02c1ed967b.MP4

해당 동영상에 대해 짧은 설명

(표에 첨부할 경우)
|            <center>동영상</center>             |    <center>설명</center>     |
| :-----------------------------------------: | :-------------------------: |
| <video src="https://user-images.githubusercontent.com/82797883/127593222-1d808786-2be1-4128-90ef-db02c1ed967b.png">| 해당 사진에 대해 짧은 설명 |
-->
|            <center>Light Theme</center>             |    <center>Dark Theme</center>     |
| :-----------------------------------------: | :-------------------------: |
| <img width="375" src="https://github.com/channel-io/BezierSwift/assets/14305293/3858ab80-06af-4781-a73d-4272ac45f009"> | <img width="375" src="https://github.com/channel-io/BezierSwift/assets/14305293/c290dc82-7a73-458c-8dbc-5328fba282a0"> |

## Reference
<!-- 기획이나 레퍼런스가 정리된 링크를 첨부합니다.  (팀챗, 노션) -->

## Issue-number
<!-- - closed: #이슈번호 -->

## Checklist
- [x] PR 제목을 라벨과 함께 명령형으로 작성했습니다.
- [x] [코딩 컨벤션](https://github.com/channel-io/ios-convention-guide) 에 맞춰서 작성했습니다
- [x] 리뷰 리퀘스트 전에 더 이상 스스로 리뷰할게 없을 정도까지 셀프 리뷰를 진행했습니다
- [ ] 변경사항에 대한 테스트코드를 추가했습니다. 또는, 테스트코드가 필요없는 이유가 있습니다(현재는 옵셔널입니다)

solchan87 committed:
[Feature] SwiftUI BezierButton 추가

solchan87 committed:
[Improvement] Configuration Action 제거

solchan87 committed:
[Improvement] Color Theme 개선 적용

solchan87 committed:
[Feature] DisabledStyle 추가

solchan87 committed:
[Improvement] BezierButton DisabledStyle 적용

solchan87 committed:
[Improvement] 불필요한 파일 제거 및 폴더 이동

solchan87 committed:
[Improvement] BezierButtonConfiguration, Equatable 준수 하도록 변경

solchan87 committed:
[Improvement] fixedSize 를 통해 Hugging 을 조절할 수 있도록 변경

solchan87 committed:
[Feature] Sample Page 추가

solchan87 committed:
[MOB-1128] Loader 추가 (#51)

* [Feature] BezierSpinnerConfiguration 추가

* [Feature] UIBezierSpinner 추가

* [Feature] BezierSpinner 추가

* [Improvement] BezierSpinner 구조 변경 및 가중치 조절

* [Etc] 코드컨벤션 적용

* [Improvement] SwiftUI Preview 추가

* [Improvement] SnapKit 적용

* [Improvement] Themeable 제거 및 Color 변경점 반영

* [Fix] Theme 변경 시 CGColor 반영

* [Etc] 코드 정리

* [Improvement] UIBezierSpinner 제거

* [Improvement] BezierSpinner -> BezierLoader 로 변경

* [Improvement] 파일 위치 변경

* [Etc] ColorScheme 제거

* [Improvement] default 반영한 init 추가 및 퀵헬프 추가

* [Etc] 오타 수정

solchan87 committed:
[Feature] Tintable 추가

solchan87 committed:
[Feature] 자주 사용하는 View Modifire 추가

solchan87 committed:
[Feature] SwiftUI Loader 를 UIKit Loader로 전환

solchan87 committed:
[Feature] Tintable 추가

solchan87 committed:
[Feature] BezierIconView 추가

solchan87 committed:
[Feature] BezierButton 구조 변경 및 Contant에 BezierIcon 추가

solchan87 committed:
[Feature] Loader Example 추가

solchan87 commented on BezierSwift/Sources/Components/Loader/UIBezierLoader.swift:
```diff
@@ -0,0 +1,148 @@
+}
+
+// MARK: - BezierSpinner
+public class UIBezierLoader: UIView {
```
SwiftUI 로 만든 Loader 에서 부모뷰 애니메이션과 중첩되는 이슈가 있어서 UIKit Loader를 래핑해서 쓰는 방향으로 변경했습니다.

solchan87 committed:
[Improvement] 누락된 ViewBuilder 추가

solchan87 commented on BezierSwift/Sources/Extensions/ViewExtensions.swift:
```diff
@@ -0,0 +1,180 @@
+  }
+  
+  @ViewBuilder
+  func scaleToFit(size: CGSize, scaleBy: ScaleDimension = .width) -> some View {
```
Loader 를 컴포넌트에 overlay 해서 사용할때 정해진 크기와 다르게 스케일로 조정할 수 있어야돼서 이 함수를 추가했습니다.
<img width="637" alt="image" src="https://github.com/user-attachments/assets/43d86cd8-9ec8-45b6-aa3f-4fecd745e684">

bonoogi commented on BezierSwift/Sources/Components/Button/BezierButton.swift:
```diff
@@ -0,0 +1,113 @@
+  // MARK: Body
+  public var body: some View {
+    Button {
+      self.action()
```
isLoading 상태에서는 인터랙션이 막히는 걸로 알고 있어요! 해당 처리가 들어가야할 것 같습니다.
https://www.notion.so/channelio/Button-95e7964c1e174aaa9923abc8ecc193b3?pvs=4#caba5cb3a60b4a579341164be1494cc1

bonoogi replied:
더불어서  가로를 꽉 채웠을 때 tertiary 이면 배경이 투명한 상태인데, 뭔가 클릭되는 영역이 배경색이 있을때랑은 다른 것 같더라구요. 가운데 부분 클릭하면 동작하기 때문에 어색하진 않은데, 이게 원래 기본적인 동작인걸까요?

https://github.com/user-attachments/assets/8ac27de7-6266-4ed3-84a0-2299a5140be7

solchan87 replied:
감사합니다. SwiftUI 에서 Alpha 0 일 경우 터치 이벤트를 뒤로 넘겨주는 롤이 있는데 이를 막기 위해서 contentShape를 추가했어야 됐는데, 이 코드에서 누락 됐었습니다. 코드 추가하고 정상 동작 확인했습니다. 감사합니다!

solchan87 replied:
@bonoogi 인터렉션은 Disabled 상태인지 단순 터치만 막는 건지 체크가 필요했어서 주디한테 문의 드렸었어요. 확인 후 Disabled 로 변경했습니다.
높이 이슈도 있었는데, fixedSize 로 했을 경우 2줄 case 를 고려하지 못해서 View 에서 isFilled 옵션을 받아서 버튼이 fit인지 fill 인지 구분하도록 했습니다. 감사합니다!
<img width="428" alt="image" src="https://github.com/user-attachments/assets/5e2e9f9d-1dbb-4713-be23-68571fbc4516">

https://desk.channel.io/root/threads/groups/(CT)BezierRedesign-244958/66ab3ca74081ce013f0e/66ab3ca74081ce013f0e

bonoogi commented on BezierSwift/Sources/Foundation/Common/Tintable.swift:
```diff
@@ -0,0 +1,47 @@
+}
+
+extension View {
+  /// 작업의 진행 단계 결정 또는 완료 시점이 확정되지 않은 경우 사용할 수 있는 요소입니다.
```
요거 BezierLoader의 설명이 일로 들어온것 같네요. BezierColor를 TintColor로 적용할 수 있다 뭐 이렇게 써도 될것같습니다.

solchan87 replied:
감사합니다! 
"추상화된 자식뷰로 TintColor를 전달하기 위한 ViewModifire입니다."
로 변경했습니다.

bonoogi commented on BezierSwift/Sources/Components/Loader/BezierLoaderConfiguration.swift:
```diff
@@ -0,0 +1,110 @@
+  }
+  
+  // MARK: Initializer
+  /// 작업의 진행 단계 결정 또는 완료 시점이 확정되지 않은 경우 사용할 수 있는 요소입니다.
+  /// - Parameters:
+  ///   - size: Loader는 small, medium 2개의 사이즈를 가질 수 있습니다.
+  ///     - medium: 화면 전체를 로딩할 때, 독립적인 요소로 쓰일 때 사용합니다.
+  ///     - small: 다른 요소와 나란히 쓰일 때 사용합니다.
+  ///       - 컴포넌트 위에 사용하는 경우에는 Icon 자리에 override 해서 사용할 것을 권장하며, 사이즈는 scale 되는 것으로 지정된 사이즈를 지키지 않아도 무방합니다.
+  ///   - variant: Loader는 primary, seondary (default), on overlay 3개의 Variant 옵션을 가질 수 있습니다.
+  ///     - primary: 화면 내에서 중요도가 높은 경우에만 사용합니다.
+  ///     - seondary(default): 보편적인 경우에 사용합니다.
+  ///     - onOverlay: dimmed layer 등과 함께할 때 사용합니다.
+  /// - Attention: 한 화면에 여러개의 Loader를 표시하지 않습니다. 이는 유저의 행동을 방해할 수 있으며, 유저에게 필요 이상의 불안정한 상태를 제공합니다.
```
```suggestion
  /// 작업의 진행 단계 결정 또는 완료 시점이 확정되지 않은 경우 사용할 수 있는 요소입니다. `variant`는 기본값인 `secondary`로 지정됩니다.
  /// - Parameters:
  ///   - size: Loader는 small, medium 2개의 사이즈를 가질 수 있습니다.
  ///     - medium: 화면 전체를 로딩할 때, 독립적인 요소로 쓰일 때 사용합니다.
  ///     - small: 다른 요소와 나란히 쓰일 때 사용합니다.
  ///       - 컴포넌트 위에 사용하는 경우에는 Icon 자리에 override 해서 사용할 것을 권장하며, 사이즈는 scale 되는 것으로 지정된 사이즈를 지키지 않아도 무방합니다.
  /// - Attention: 한 화면에 여러개의 Loader를 표시하지 않습니다. 이는 유저의 행동을 방해할 수 있으며, 유저에게 필요 이상의 불안정한 상태를 제공합니다.
```

solchan87 replied:
감사합니다. 베지어 버튼쪽도 비슷한 방향으로 수정했습니다.

solchan87 committed:
[Improvement] 잘못된 주석 수정

solchan87 committed:
[Improvement] Default Value를 포함한 Initializer 주석 수정

solchan87 committed:
[Feature] visible ViewModifire 추가

solchan87 committed:
[Feature] isFilled 옵션 추가

[description]
처음에 작성했던 FixedSize 는 화면 범위를 넘어가는 경우 max 사이즈 범위를 잡기 어려워서 내부적으로 Width를 조정할 수 있게 isFilled 라는 옵션을 추가했습니다.

solchan87 reviewed (commented):
핀 리뷰 감사합니다!

solchan87 committed:
[Fix] 줄수 무제한으로 Height -> Vertical Padding 으로 변경

solchan87 committed:
[Etc] Pressed Color 관련 주석 추가

bonoogi reviewed (approved):
따봉!

jam0128 commented on BezierSwift/Sources/Components/Button/BezierButton.swift:
```diff
@@ -0,0 +1,137 @@
+    }
+    .buttonStyle(BezierButtonStyle(configuration: self.configuration))
+    .applyDisabledStyle()
+    .disabled(self.isLoading)
```
베지어 버튼의 disabled가 isLoading에 의해서만 결정되진않을 것 같습니다. disable 되는 여러가지 조건중에 로딩이 있는거아닐까요? 잘못이해한거면 알려주세용!

solchan87 replied:
네네 .disabled(false) 의 경우 뒤에 오는 disabled 에 영향을 주지 않아서 조건별로 사용이 가능합니다. 문서상 disabled 조건으로 isLoading 하나가 있어서 내부에 명시해줬습니다.

<img width="238" alt="image" src="https://github.com/user-attachments/assets/e5cd90fc-67f5-4807-aa19-2730c2bd93b7">

jam0128 commented on BezierSwift/Sources/Foundation/Common/Tintable.swift:
```diff
@@ -0,0 +1,47 @@
+  }
+}
+
+private struct TintableView<Content: View>: View {
```
저 이 tintable 을 정확히 이해가 안되긴해서 설명한번 해주실 수 있나용? 자식뷰로 들어갈 때 마다 계속 붙여줘야하는거면, 그냥 하나씩 포그라운드 주는거랑 어떤차이점과 이점이 있는지 감이 잘 안와서요 ㅠㅠ

solchan87 replied:
TintableView 로 따로 만든 이유는 내부에 들어가는 slot 별로 적용 범위가 달라서(emoji: X, icon: O, avatar: X) environment를 통해서 옵션 형태로 선언한 부분에서말 쓸 수 있도록 구현했습니다.

jam0128 commented:
+로 댓글에 남겼는데, 
<img width="470" alt="image" src="https://github.com/user-attachments/assets/528bbae2-5e7f-4501-920f-8c1372baf670">

solchan87 commented:
Bezier...Configuration 에는 문서상 정의된 Variants 들만 모아둔 곳입니다. isFilled는 OS별로 구현 방법이 달라서 Bezier...Configuration 에서 빠진 파라미터로 구현했습니다.

jam0128 reviewed (approved):
구두논의한 부분들있어서어푸룹해둘게요!

solchan87 committed:
[Improvement] Loader 스펙 추가

[description]
로더 스펙이 추가 되었습니다.
노션: https://www.notion.so/channelio/Loader-e1e30e3fdcb14a3ba94b900d24e0d58e
팀챗: https://desk.channel.io/root/threads/groups/(CT)BezierRedesign-244958/66a9c7f6c49bd5d51d75/66a9c7f6c49bd5d51d75

solchan87 committed:
[Feature] BezierButton 스펙 변경

[description]
슬롯에 대한 논의가 있었고, 모든 UI 가 들어올 수 있도록 열어주지만, 디자인 상으로 가이드가 있을 수 있다로 결정 되었습니다.
관련하여 연관 프로퍼티들을 모두 제거합니다.
